test_util/src/pty.rs

```diff
@@ -0,0 +1,437 @@
+          let console_handle = WinHandle::new(console_handle);
+          move || {
+            WaitForSingleObject(thread_handle.as_raw(), INFINITE);
+            // wait for the reading thread to catch up
+            std::thread::sleep(Duration::from_millis(200));
+            // close the console handle which will close the
+            // stdout pipe for the reader
+            ClosePseudoConsole(console_handle.take());
+          }
+        });
+
+        Self {
+          stdin_write_handle,
+          stdout_read_handle,
+          _process_handle: WinHandle::new(proc_info.hProcess),
+          _thread_handle: thread_handle,
+          _attribute_list: attribute_list,
+        }
+      }
+    }
+  }
+
+  impl Read for WinPseudoConsole {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+      unsafe {
+        let mut bytes_read = 0;
+        // don't bother checking the result of this call as it will be false
+        // when the stdout pipe is closed and bytes_read will be 0 in that case
+        ReadFile(
+          self.stdout_read_handle.as_raw(),
+          buf.as_mut_ptr() as _,
+          buf.len() as u32,
+          &mut bytes_read,
+          ptr::null_mut(),
+        );
+        Ok(bytes_read as usize)
+      }
+    }
+  }
+
+  impl Pty for WinPseudoConsole {
+    fn write_text(&mut self, text: &str) {
+      // windows psuedo console requires a \r\n to do a newline
+      let newline_re = regex::Regex::new("\r?\n").unwrap();
+      self
+        .write_all(newline_re.replace_all(text, "\r\n").as_bytes())
+        .unwrap();
+    }
+  }
+
+  impl std::io::Write for WinPseudoConsole {
+    fn write(&mut self, buffer: &[u8]) -> std::io::Result<usize> {
+      unsafe {
+        let mut bytes_written = 0;
+        assert_win_success!(WriteFile(
+          self.stdin_write_handle.as_raw(),
+          buffer.as_ptr() as *const _,
+          buffer.len() as u32,
+          &mut bytes_written,
+          ptr::null_mut(),
+        ));
+        Ok(bytes_written as usize)
+      }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+      Ok(())
+    }
+  }
+
+  struct WinHandle {
+    inner: HANDLE,
+    close_on_drop: bool,
+  }
+
+  impl WinHandle {
+    pub fn new(handle: HANDLE) -> Self {
+      WinHandle {
+        inner: handle,
+        close_on_drop: true,
+      }
+    }
+
+    pub fn duplicate(&self) -> WinHandle {
+      unsafe {
+        let process_handle = GetCurrentProcess();
+        let mut duplicate_handle = ptr::null_mut();
+        assert_win_success!(DuplicateHandle(
+          process_handle,
+          self.inner,
+          process_handle,
+          &mut duplicate_handle,
+          0,
+          0,
+          DUPLICATE_SAME_ACCESS,
+        ));
+
+        WinHandle::new(duplicate_handle)
+      }
+    }
+
+    pub fn as_raw(&self) -> HANDLE {
+      self.inner
+    }
+
+    pub fn take(mut self) -> HANDLE {
+      self.close_on_drop = false;
```

Comment:
Given that the only purpose of `close_on_drop` is to make `take()` work, I'd suggest this solution instead:

```rust
let handle = self.inner;
std::mem::forget(self);
handle
```

Reply:
Thanks! Much better. I had a brief moment of trying to remember what this function was called, but I... forgot.